jserrors.go

```diff
@@ -0,0 +1,185 @@
+
+	// ErrNoMatchingStream is returned when stream lookup by subject is unsuccessful.
+	ErrNoMatchingStream JetStreamError = &jsError{message: "no stream matches subject"}
+
+	// ErrSubjectMismatch is returned when the provided subject does not match consumer's filter subject.
+	ErrSubjectMismatch JetStreamError = &jsError{message: "subject does not match consumer"}
+
+	// ErrContextAndTimeout is returned when attempting to use both context and timeout.
+	ErrContextAndTimeout JetStreamError = &jsError{message: "context and timeout can not both be set"}
+
+	// ErrCantAckIfConsumerAckNone is returned when attempting to ack a message for consumer with AckNone policy set.
+	ErrCantAckIfConsumerAckNone JetStreamError = &jsError{message: "cannot acknowledge a message for a consumer with AckNone policy"}
+
+	// DEPRECATED: ErrInvalidDurableName is no longer returned and will be removed in future releases
+	// Use ErrInvalidConsumerName instead
+	ErrInvalidDurableName = errors.New("nats: invalid durable name")
+)
+
+// Error code represents JetStream error codes returned by the API
+type ErrorCode uint16
+
+const (
+	JSErrCodeJetStreamNotEnabledForAccount ErrorCode = 10039
+	JSErrCodeJetStreamNotEnabled           ErrorCode = 10076
+
+	JSErrCodeStreamNotFound  ErrorCode = 10059
+	JSErrCodeStreamNameInUse ErrorCode = 10058
+
+	JSErrCodeConsumerNotFound      ErrorCode = 10014
+	JSErrCodeConsumerNameExists    ErrorCode = 10013
+	JSErrCodeConsumerAlreadyExists ErrorCode = 10105
+
+	JSErrCodeMessageNotFound ErrorCode = 10037
+
+	JSErrCodeBadRequest ErrorCode = 10003
+)
+
+// APIError is included in all API responses if there was an error.
+type APIError struct {
+	Code        int       `json:"code"`
+	ErrorCode   ErrorCode `json:"err_code"`
+	Description string    `json:"description,omitempty"`
+}
+
+// Error prints the JetStream API error code and description
+func (e *APIError) Error() string {
+	return fmt.Sprintf("nats: API error %d: %s", e.ErrorCode, e.Description)
+}
+
+// APIError implements the JetStreamError interface.
+func (e *APIError) APIError() *APIError {
+	return e
+}
+
+// Is matches against an APIError.
+func (e *APIError) Is(err error) bool {
+	if e == nil {
+		return false
+	}
+	// Extract internal APIError to match against.
+	var aerr *APIError
+	ok := errors.As(err, &aerr)
+	if !ok {
+		return ok
+	}
+	return e.ErrorCode == aerr.ErrorCode
+}
+
+// JetStreamError is an error result that happens when using JetStream.
+// In case of client-side error, `APIError()` returns nil
+type JetStreamError interface {
+	APIError() *APIError
+	error
+}
+
+type jsError struct {
+	apiErr  *APIError
+	message string
+}
+
+func (err *jsError) APIError() *APIError {
+	return err.apiErr
+}
+
+func (err *jsError) Error() string {
+	if err.apiErr != nil && err.apiErr.Description != "" {
+		return err.apiErr.Error()
+	}
+	return fmt.Sprintf("nats: %s", err.message)
+}
+
+func (err *jsError) Unwrap() error {
```

Comment:
More info about features of wrapping and `Unwrap` can be found here: 

- https://pkg.go.dev/errors#pkg-functions
- https://github.com/golang/go/wiki/ErrorValueFAQ